server/src/main/java/org/opensearch/gateway/AsyncShardFetch.java

```diff
@@ -77,18 +76,22 @@ public abstract class AsyncShardFetch<T extends BaseNodeResponse> implements Rel
      * An action that lists the relevant shard data that needs to be fetched.
      */
     public interface Lister<NodesResponse extends BaseNodesResponse<NodeResponse>, NodeResponse extends BaseNodeResponse> {
-        void list(ShardId shardId, @Nullable String customDataPath, DiscoveryNode[] nodes, ActionListener<NodesResponse> listener);
+        void list(Map<ShardId, String> shardIdsWithCustomDataPath, DiscoveryNode[] nodes, ActionListener<NodesResponse> listener);
+
     }
 
     protected final Logger logger;
     protected final String type;
-    protected final ShardId shardId;
-    protected final String customDataPath;
+
+    protected final Map<ShardId, String> shardToCustomDataPath;
```

Comment:
We can avoid these maps as they will end up taking additional space for ShardId keys. Lets see if we can use an encapsulation instead.

Reply:
If we need to use ShardID type of keys then it is effectively same thing. Can you mention the reason why we need to use it as a key?